## Vector Embedding

In [1]:
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("all-MiniLM-L6-v2")

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.5k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [ ]:
# Turning the query into a vector
q1 = "Can I still join the course after the start date?"
v1 = model.encode(q1)
v1

array([ 2.13903822e-02, -7.39799738e-02,  1.42068532e-03,  2.13816594e-02,
        2.45113522e-02,  3.15582715e-02, -1.10839754e-01, -1.05017476e-01,
       -6.18259311e-02, -6.42313249e-03,  3.72394337e-03,  9.06393453e-02,
       -9.49937291e-03,  6.53976649e-02,  1.10946922e-02, -2.10097264e-02,
       -3.35125625e-02, -4.31677215e-02,  9.96347517e-03,  1.41970059e-02,
       -6.40415549e-02, -7.04177516e-03, -7.91187733e-02,  5.80030791e-02,
        1.30209117e-03,  4.19735443e-03,  5.70979342e-02,  6.39447942e-02,
        2.49902811e-02, -3.95876542e-02, -3.79506089e-02,  2.70394497e-02,
        1.79423504e-02,  1.72272343e-02,  3.43311168e-02,  9.29055363e-03,
        5.86055331e-02, -4.97789979e-02, -5.05368831e-03,  4.34328318e-02,
       -1.56623088e-02, -2.97534410e-02, -5.13327355e-03,  5.13414964e-02,
        6.16064016e-03,  6.86980635e-02, -1.29505526e-02, -5.61938323e-02,
       -1.08265067e-02,  5.96683919e-02,  5.29940054e-02, -3.42755020e-02,
       -4.15274389e-02, -

In [17]:
len(v1)

384

v1 is a vector, an array of 384 numbers. Each number stands for some concept the model learned. We can't read off what any one of them means. But two vectors with similar values point to texts about similar things.

In [3]:
# Encoding the document
d  = "You don't need to register. You're accepted. You can also just start learning and submitting homework without registering."
dv = model.encode(d)
dv

array([-4.81206961e-02, -1.29942089e-01, -4.86808410e-03,  1.37495250e-02,
       -8.51094816e-03,  6.97840974e-02, -7.55662657e-03,  4.09571454e-02,
       -1.02165624e-01, -3.45050208e-02,  2.85521224e-02, -5.81430532e-02,
       -1.85113400e-02, -3.17839086e-02,  5.29631525e-02,  3.96292806e-02,
       -9.84592643e-03, -6.45324513e-02,  7.05097616e-02, -6.39589056e-02,
       -3.55156884e-02, -1.75489970e-02, -3.58401686e-02,  5.50822238e-04,
        3.38445418e-02,  2.91763060e-02,  3.24278586e-02, -5.64803602e-03,
        8.42592716e-02,  5.90306148e-02,  2.08772700e-02, -5.44834044e-03,
        2.98403576e-02,  1.30820340e-02,  6.36077896e-02, -3.06394473e-02,
        6.28460273e-02, -1.69217978e-02, -4.65758443e-02, -5.13762385e-02,
       -3.81560437e-02, -7.01197237e-02, -9.55517665e-02,  5.01640327e-02,
        8.57910048e-03,  1.19207231e-02, -1.40597336e-02, -7.98985083e-03,
       -3.44238654e-02, -2.34269891e-02,  2.89466120e-02,  2.68904828e-02,
       -4.87409197e-02, -

In [16]:
len(dv)

384

In [4]:
# Next, we compare the query against the document using dot product:
v1.dot(dv)

np.float32(0.32332408)

In [6]:
# Trying an unrelated query
q2 = "How to install Docker on Windows?"
v2 = model.encode(q2)
v2

array([-2.35678032e-02,  7.85555989e-02,  2.33792141e-02, -6.72473432e-03,
       -2.58875005e-02,  1.18257152e-02,  5.29537385e-04,  9.57368035e-03,
        6.23929314e-02,  2.91426964e-02, -4.18229811e-02, -6.62150159e-02,
       -2.76613981e-02,  5.41518629e-02, -1.92782115e-02, -2.46782005e-02,
       -2.89789750e-03,  8.40461701e-02,  1.03637436e-02, -6.35400368e-03,
       -1.20814092e-01,  1.80162936e-02, -5.70900925e-02, -7.92970806e-02,
       -9.25548654e-03, -3.24045792e-02, -1.95663199e-02, -2.65411008e-02,
       -9.41327121e-03,  4.31234874e-02, -2.08750684e-02, -7.86434188e-02,
        9.19373631e-02, -3.75259039e-03,  5.13070859e-02, -5.73641574e-03,
        5.61954491e-02,  3.73760704e-03, -6.88236132e-02, -4.57000099e-02,
        2.83075906e-02, -3.49764638e-02, -5.55249746e-04, -3.09770145e-02,
        4.34275456e-02, -4.48056422e-02,  6.84110001e-02,  4.14165406e-04,
        5.82685806e-02,  6.66982755e-02, -8.95876139e-02, -2.69941781e-02,
        1.06747057e-02, -

In [7]:
v2.dot(dv)

np.float32(0.019730462)

The first query (asking about registration) (.32) is higher, therefore this query is more similar to the document about registration compared to second query about docker (.01). The dot product tells us how similar the document is to the query in question.

## Embedding a Dataset 

In [ ]:
# Loading data source
from ingest import load_faq_data

documents = load_faq_data()

In [12]:
# question and answer will be embedded both together so that a query can match against both the question text and answer text in the index.
texts = []

for doc in documents:
    text = doc["question"] + " " + doc["answer"]
    texts.append(text)

In [13]:
# Next step is to generate embeddings
# We need to embed in batches as processing can take a long time
# import tqdm to review process
from tqdm.auto import tqdm

In [14]:
# Now chunk the dataset into batches and encode each batch
batch_size = 50
vectors = []

for i in tqdm(range(0, len(texts), batch_size)):
    batch = texts[i:i + batch_size]
    batch_vectors = model.encode(batch)
    vectors.extend(batch_vectors)

len(vectors)

  0%|          | 0/27 [00:00<?, ?it/s]

1350

In [ ]:
# Next we turn it into a 2-D array (matrix) where row = docs and cols = dimension of the vector which represent clues or meanings of the text the model has learned and is represented by a number
import numpy as np
X = np.array(vectors)
X

array([[-0.02670616, -0.12245759,  0.01594421, ..., -0.0023065 ,
        -0.11218402, -0.02365559],
       [-0.01099559, -0.11074748, -0.02536943, ...,  0.09022236,
        -0.02697359,  0.01975664],
       [-0.08896557, -0.06128179,  0.00775611, ...,  0.04059712,
         0.00479282, -0.02745942],
       ...,
       [-0.03652923,  0.0141543 , -0.06838647, ...,  0.04316789,
         0.08105531, -0.02148631],
       [-0.1309159 , -0.06990598, -0.0093188 , ..., -0.00044342,
        -0.01285736,  0.01426917],
       [-0.07984782,  0.01926984,  0.02544974, ..., -0.03368021,
        -0.01884026,  0.05837055]], shape=(1350, 384), dtype=float32)

In [18]:
len(X)

1350

## Scoring the Documents

In [19]:
# First we embed the query
query = "Can I still join the course after the start date?"
v_query = model.encode(query)

In [25]:
print(f"Shape: {v_query.shape},\n View of array: {v_query}")

Shape: (384,),
 View of array: [ 2.13903822e-02 -7.39799738e-02  1.42068532e-03  2.13816594e-02
  2.45113522e-02  3.15582715e-02 -1.10839754e-01 -1.05017476e-01
 -6.18259311e-02 -6.42313249e-03  3.72394337e-03  9.06393453e-02
 -9.49937291e-03  6.53976649e-02  1.10946922e-02 -2.10097264e-02
 -3.35125625e-02 -4.31677215e-02  9.96347517e-03  1.41970059e-02
 -6.40415549e-02 -7.04177516e-03 -7.91187733e-02  5.80030791e-02
  1.30209117e-03  4.19735443e-03  5.70979342e-02  6.39447942e-02
  2.49902811e-02 -3.95876542e-02 -3.79506089e-02  2.70394497e-02
  1.79423504e-02  1.72272343e-02  3.43311168e-02  9.29055363e-03
  5.86055331e-02 -4.97789979e-02 -5.05368831e-03  4.34328318e-02
 -1.56623088e-02 -2.97534410e-02 -5.13327355e-03  5.13414964e-02
  6.16064016e-03  6.86980635e-02 -1.29505526e-02 -5.61938323e-02
 -1.08265067e-02  5.96683919e-02  5.29940054e-02 -3.42755020e-02
 -4.15274389e-02 -5.43295592e-02 -7.14877341e-03  1.28990695e-01
  2.09546387e-02  1.25907371e-02 -1.04924925e-01  1.5725091

In [28]:
# Compute dot product against all docs
scores = X.dot(v_query)
print(f"Shape:{scores.shape} \n\n View of output: {scores}")

Shape:(1350,) 

 View of output: [ 0.4874057   0.20991942  0.762941   ... -0.08637965  0.03759797
 -0.03037042]


Could be done with for loop but X.dot(v_query) is much faster. This results in one score per document.
```
scores = [v_query.dot(X[i]) for i in range(len(X))]
```

In [29]:
# Finding the best match; The highest score is the most similar document
idx = np.argmax(scores)
idx, scores[idx]

(np.int64(2), np.float32(0.762941))

This returns document 2 with score 0.76.

In [31]:
# reviewing document
documents[idx]

{'id': '3f1424af17',
 'course': 'data-engineering-zoomcamp',
 'section': 'General Course-Related Questions',
 'question': 'Course: Can I still join the course after the start date?',
 'answer': "Yes, even if you don't register, you're still eligible to submit the homework.\n\nBe aware, however, that there will be deadlines for turning in homeworks and the final projects. So don't leave everything for the last minute."}

## Pulling Top 5 Results
np.argsort sorts from lowest to highest, so the last 5 are the top ones:

In [32]:
top5 = np.argsort(scores)[-5:]
top5

array([  7, 538, 907, 625,   2])

In [33]:
# Next we reverse the order so it's highest to lowest in terms of similarity
top5 = top5[::-1]
top5

array([  2, 625, 907, 538,   7])

In [34]:
scores[top5]

array([0.762941  , 0.7579372 , 0.7192133 , 0.65363127, 0.5601001 ],
      dtype=float32)

In [35]:
#  Alternative: We negate the scores first, so the largest becomes the smallest. Then argsort puts the best matches at the front.
top5 = np.argsort(-scores)[:5]
top5

array([  2, 625, 907, 538,   7])

In [36]:
# Reviewing the top 5 docs
for idx in top5:
    print(scores[idx])
    print(documents[idx])
    print()

0.762941
{'id': '3f1424af17', 'course': 'data-engineering-zoomcamp', 'section': 'General Course-Related Questions', 'question': 'Course: Can I still join the course after the start date?', 'answer': "Yes, even if you don't register, you're still eligible to submit the homework.\n\nBe aware, however, that there will be deadlines for turning in homeworks and the final projects. So don't leave everything for the last minute."}

0.7579372
{'id': '2d8b16c2a0', 'course': 'mlops-zoomcamp', 'section': 'General Course-Related Questions', 'question': 'Course - Can I still join the course after the start date?', 'answer': "Yes, even if you don't register, you're still eligible to submit the homeworks as long as the form is still open and accepting submissions.\n\nBe aware, however, that there will be deadlines for turning in the final projects. So don't leave everything to the last minute."}

0.7192133
{'id': '41aabbd7c5', 'course': 'machine-learning-zoomcamp', 'section': 'General Course-Related 

This is vector search in its simplest form. We embed the query, compute dot products against all documents, and return the highest-scoring ones.